Add IQS028 rule (#2393)
home-assistant/developers.home-assistant: docs/core/integration-quality-scale/rules/iqs028.md

---

> title: "IQS028 - The integration support passing in a websession"

Reviewer: ```suggestion
title: "IQS028 - The integration dependency supports passing in a websession"
```

---

IQS028 is as follows:

> The integration support passing in a websession

Reviewer: ```suggestion
The integration dependency supports passing in a websession
```

## Reasoning

Since many devices and services are connected via HTTP, the number of active web sessions can be high.

> To improve the efficiency of those web sessions, it is recommended to support passing in a web session.

Reviewer: ```suggestion
To improve the efficiency of those web sessions, it is recommended to support passing in a web session to the dependency client that is used by the integration.
```

> These sessions are tweaked towards being used in Home Assistant and is more efficient.

Reviewer: It's more that using the same session enables connection pooling.
```suggestion
```

> Since a lot of devices and services are connected to via HTTP, the amount of websessions that are active can be a lot.
> To improve the efficiency of those websessions, it is recommended to support passing in a websession.
> These sessions are tweaked towards being used in Home Assistant and is more efficient.

Reviewer 2: ```suggestion
Since many devices and services are connected via HTTP, the number of active web sessions can be high.
To improve the efficiency of those web sessions, it is recommended to support passing in a web session.
These sessions are tweaked towards being used in Home Assistant and is more efficient.
```

> Home Assistants supports this for `aiohttp` and `httpx`.

Reviewer 2: Let's link them up 🚀 

```suggestion
Home Assistants supports this for [`aiohttp`](https://docs.aiohttp.org/en/stable/) and [`httpx`](https://www.python-httpx.org/).
```

> Home Assistants supports this for [`aiohttp`](https://docs.aiohttp.org/en/stable/) and [`httpx`](https://www.python-httpx.org/).

Reviewer: ```suggestion
Home Assistants supports this for [`aiohttp`](https://docs.aiohttp.org/en/stable/) and [`httpx`](https://www.python-httpx.org/). This means that the integration dependency should use either of those two libraries.
```

## Example implementation

> In the example below, a `aiohttp` session is passed in to the client.

Reviewer: ```suggestion
In the example below, an `aiohttp` session is passed in to the client.
```

The equivalent for `httpx` would be `get_async_client`.

```python
async def async_setup_entry(hass: HomeAssistant, entry: MyConfigEntry) -> bool:
    """Set up my integration from a config entry."""

    client = MyClient(entry.data[CONF_HOST], async_get_clientsession(hass))
```

:::info
There are cases where you might not want a shared session, for example when cookies are used.
In that case, you can create a new session using `async_create_clientsession` for `aiohttp` and `create_async_httpx_client` for `httpx`.
:::

## Additional resources

## Exceptions

> There are no exceptions to this rule.

Author: Do we know any exceptions to this?

Reviewer 2: Yup, in case `aiohttp` and `httpx` aren't used? :)

Author: Okay so to combine this with the async dependency rule.

If someone created a library that worked with requests, and uses the `hass.loop` as parameter to create their own executor jobs in the library to make the API we use async. (I think this is a realistic use case).

Do we still allow them for both rules? Because in that case the dependency is "async" (not optimal async, but we talk async to the library) thus that rule would be checked, and this rule would not be checked since its exempt.

Reviewer: We shouldn't allow such a library, I think. That's a very bad pattern.

Author: I agree. But what is the best way to enforce that? In what case would one be making HTTP calls in async context and not be using `aiohttp` or `httpx`

Reviewer: To meet this rule, the integration should use an async library that either uses aiohttp or httpx for http requests. I don't think we need an exception for this.

If the library doesn't do http requests the rule doesn't apply.

Reviewer 2: That would be the exception I guess; If the integration doesn't do HTTP requests.

Author: Then we should move the exception from the title to the exceptions
